src/catalog/src/remote/manager.rs

```diff
@@ -784,89 +743,119 @@ impl SchemaProvider for RemoteSchemaProvider {
+            "Successfully set catalog table entry, key: {}, table value: {:?}",
+            table_key, table_value
+        );
 
-        let prev = std::thread::spawn(move || {
-            common_runtime::block_on_read(async move {
-                let _guard = mutex.lock(table_key.clone()).await;
-                backend
-                    .set(
-                        table_key.as_bytes(),
-                        &table_value.as_bytes().context(InvalidCatalogValueSnafu)?,
-                    )
-                    .await?;
-                debug!(
-                    "Successfully set catalog table entry, key: {}, table value: {:?}",
-                    table_key, table_value
-                );
-
-                let tables = tables.load();
-                let prev = tables.insert(name, table);
-                Ok(prev)
-            })
-        })
-        .join()
-        .unwrap();
-        prev
+        // TODO(hl): retrieve prev table info using cas
+        Ok(None)
     }
 
-    fn rename_table(&self, _name: &str, _new_name: String) -> Result<TableRef> {
+    async fn rename_table(&self, _name: &str, _new_name: String) -> Result<TableRef> {
         UnimplementedSnafu {
             operation: "rename table",
         }
         .fail()
     }
 
-    fn deregister_table(&self, name: &str) -> Result<Option<TableRef>> {
-        let table_name = name.to_string();
-        let table_key = self.build_regional_table_key(&table_name).to_string();
-        let backend = self.backend.clone();
-        let mutex = self.mutex.clone();
-        let tables = self.tables.clone();
-        let prev = std::thread::spawn(move || {
-            common_runtime::block_on_read(async move {
-                let _guard = mutex.lock(table_key.clone()).await;
-                backend.delete(table_key.as_bytes()).await?;
-                debug!(
-                    "Successfully deleted catalog table entry, key: {}",
-                    table_key
-                );
+    async fn deregister_table(&self, name: &str) -> Result<Option<TableRef>> {
+        let table_key = self.build_regional_table_key(name).to_string();
+        self.backend.delete(table_key.as_bytes()).await?;
+        debug!(
+            "Successfully deleted catalog table entry, key: {}",
+            table_key
+        );
 
-                let tables = tables.load();
-                let prev = tables.remove(&table_name).map(|en| en.1);
-                Ok(prev)
-            })
-        })
-        .join()
-        .unwrap();
-        prev
+        let reference = TableReference {
+            catalog: &self.catalog_name,
+            schema: &self.schema_name,
+            table: name,
+        };
+        // deregistering table does not necessarily mean dropping the table
+        let table = self
+            .engine_manager
+            .engine(MITO_ENGINE)
+            .context(TableEngineNotFoundSnafu {
+                engine_name: MITO_ENGINE,
+            })?
```

Comment:
Maybe we can add an engine field in the  `DeregisterTableRequest` or fetch table info again? I guess it's safe to do this, we don't support changing the table engine after a table is created.

Reply:
You are right, I added a commit to fix this by retrieving regional table value and get the table engine name to find the table engine.